File: samples/golang/ftl/tibMapGet/tibMapGet.go
package main

import (
	"flag"
	"fmt"
	"log"
	"os"
	"strings"

	"tibco.com/ftl"
)

var (
	appName            string = "tibmapget"
	realmServer        string = "http://localhost:8080"
	mapEndpoint        string = "tibmapget-endpoint"
	trcLevel           string
	user               string = "guest"
	password           string = "guest-pw"
	identifier         string
	secondary          string
	mapName            string = "tibmap"
	DEFAULT_URL        string = "http://localhost:8080"
	DEFAULT_SECURE_URL string = "https://localhost:8080"
	keyList            string
	clientLabel        string
	lockName           string = "lock-tibMapGet"

	done bool = false

	realm   ftl.Realm
	tibMap  ftl.Map
	msg     ftl.Message
	mapIter ftl.MapIterator

	count     int    = 0
	trustAll  bool   = false
	trustFile string = ""
	removeMap bool   = false
	help      bool   = false
)

var Usage = func() {
	fmt.Fprintf(os.Stderr, "Usage:  tibMapGet [options] <realmserver_url>\n")
	fmt.Fprintf(os.Stderr, "Default url is %s\n", realmServer)
	fmt.Fprintf(os.Stderr, "      -a,  --application <name>\n"+
		"         --client-label <label>\n"+
		"    -e,  --endpoint <name>\n"+
		"    -h,  --help\n"+
		"    -id, --identifier <name>\n"+
		"    -l,  --keyList <string>[,<string>]\n"+
		"            if not provided the entire key space will be iterated\n"+
		"    -n,  --mapname <name>\n"+
		"    -p,  --password <string>\n"+
		"    -u,  --user <string>\n"+
		"    -r   --removemap \n"+
		"            the mapname command line arg needs to be specified as well\n"+
		"    -s,  --secondary <string>\n"+
		"    -t,  --trace <level>\n"+
		"            where <level> can be:\n"+
		"                  off, severe, warn, info, debug, verbose\n"+
		"         --trustall\n"+
		"         --trustfile <file>\n")
}

func parseArgs() {
	flag.StringVar(&appName, "a", "tibmapget", "Application Name")
	flag.StringVar(&appName, "application", "tibmapget", "Application Name")

	flag.StringVar(&clientLabel, "client-label", "", "client label")
	flag.StringVar(&clientLabel, "cl", "", "client label")

	flag.StringVar(&mapEndpoint, "endpoint", "tibmapget-endpoint", "Endpoint Name")
	flag.StringVar(&mapEndpoint, "e", "tibmapget-endpoint", "Endpoint Name")

	flag.StringVar(&identifier, "id", "", "Identifier")
	flag.StringVar(&identifier, "identifier", "", "Identifier")

	flag.StringVar(&keyList, "keyList", "", "Key List")
	flag.StringVar(&keyList, "l", "", "Key List")

	flag.StringVar(&mapName, "mapname", "tibmap", "Map Name")
	flag.StringVar(&mapName, "n", "tibmap", "Map Name")

	flag.StringVar(&password, "p", "", "Password for realm server")
	flag.StringVar(&password, "password", "", "Password for realm server")

	flag.BoolVar(&removeMap, "removemap", false, "Remove Map")

	flag.StringVar(&secondary, "secondary", "", "Secondary")
	flag.StringVar(&secondary, "s", "", "Secondary")

	flag.StringVar(&trcLevel, "t", "", "Trace level")
	flag.StringVar(&trcLevel, "trace", "", "Trace level")

	flag.BoolVar(&trustAll, "trustall", false, "Trust all secure realm servers")

	flag.StringVar(&trustFile, "trustfile", "", "Trust realm servers with this certificate")

	flag.StringVar(&user, "u", "", "User name for realm server")
	flag.StringVar(&user, "user", "", "User name for realm server")

	flag.BoolVar(&help, "help", false, "print usage")
	flag.BoolVar(&help, "h", false, "print usage")

	flag.Parse()

	if help {
		Usage()
		os.Exit(0)
	}

	if flag.NArg() > 0 {
		realmServer = flag.Arg(0)
	}
}

// go routine to handle notifications
func processNotifications(realm ftl.Realm) {
	notificationsPipe := make(chan ftl.Notification)
	realm.SetNotificatonChannel(notificationsPipe)

	for notification := range notificationsPipe {
		if notification.Notify == ftl.NotificationClientDisabled {
			log.Println("application administratively disabled: " + notification.Reason)
			done = true
		} else {
			log.Printf("Notification type is %d : %s", notification.Notify, notification.Reason)
			done = true
		}
	}
}

func work() int {
	if trcLevel != "" {
		ftl.SetLogLevel(trcLevel)
	}

	// Set up properties of realm connection
	realmProps := make(ftl.Props)
	if user != "" {
		realmProps[ftl.RealmPropertyStringUsername] = user
		realmProps[ftl.RealmPropertyStringUserpassword] = password
	}
	if identifier != "" {
		realmProps[ftl.RealmPropertyStringAppinstanceIdentifier] = identifier
	}
	if secondary != "" {
		realmProps[ftl.RealmPropertyStringSecondaryServer] = secondary
	}
	if trustFile != "" {
		realmProps[ftl.RealmPropertyLongTrustType] = ftl.RealmHTTPSConnectionUseSpecifiedTrustFile
		realmProps[ftl.RealmPropertyStringTrustFile] = trustFile
	} else if trustAll {
		realmProps[ftl.RealmPropertyLongTrustType] = ftl.RealmHTTPSConnectionTrustEveryone
	}
	if clientLabel != "" {
		realmProps[ftl.RealmPropertyStringClientLabel] = clientLabel
	}

	// Connect to the realm server
	realm, err := ftl.Connect(realmServer, appName, realmProps)
	if err != nil {
		log.Fatal(ftl.ErrStack(err))
	}
	defer realm.Close()

	go processNotifications(realm)

	tibMap, _ := realm.NewMap(mapEndpoint, mapName, nil)
	if err != nil {
		log.Fatal(ftl.ErrStack(err))
	}
	defer tibMap.Close()

	lock, err := realm.NewLock(lockName, nil)
	if err != nil {
		log.Fatal(ftl.ErrStack(err))
	}

	if keyList != "" {
		keys := strings.Split(keyList, ",")
		for _, key := range keys {
			fmt.Printf("getting value for key %s\n", key)
			msg, err = tibMap.Get(key, lock)
			if err != nil {
				log.Fatal(ftl.ErrStack(err))
			}
			msg.String()
			if msg.String() != "" {
				fmt.Printf("Key = %s, value = %s", key, msg.String())
			} else {
				fmt.Printf("Currently no key %s in the map", key)
			}
		}
	} else {
		fmt.Printf("Iterating keys is the map\n")

		mapIter, err := tibMap.NewIterator(lock, nil)
		if err != nil {
			log.Fatal(ftl.ErrStack(err))
		}
		defer mapIter.Destroy()

		for {
			iterNext, err := mapIter.Next()
			if err != nil {
				log.Fatal(ftl.ErrStack(err))
			}
			if iterNext == false {
				break
			}
			key, _ := mapIter.CurrentKey()
			msg, _ := mapIter.CurrentValue()
			if key != "" {
				fmt.Fprintf(os.Stderr, "Key: %s Value: %s\n", key, msg.String())
			}
		}
	}

	if mapName != "" && removeMap {
		fmt.Printf("Removing map %s", mapName)
		realm.RemoveMap(mapEndpoint, mapName, nil)
	}

	return 0
}

func main() {
	// print the banner
	log.Println("--------------------------------------------------------------")
	log.Println("TibMapGet")
	log.Println(ftl.Version())
	log.Println("--------------------------------------------------------------")
	parseArgs()
	work()
}

File: samples/java/ftl/TibMapGet.java
/*
 * Copyright (c) 2010-2017 TIBCO Software Inc.
 * Licensed under a BSD-style license. Refer to [LICENSE]
 * For more information, please contact:
 * TIBCO Software Inc., Palo Alto, California, USA
 */

/*
 * This is a sample of a basic java program which demonstrates the TibMap API that
 * gets the requested number of keys and their values that are stored in 'tibmap' at the 
 * persistence server. If keyList is not specified then it iterates over all the keys in 
 * the tibmap.
 */

package com.tibco.ftl.samples;

import java.util.*;
import com.tibco.ftl.*;
import com.tibco.ftl.exceptions.*;

public class TibMapGet implements NotificationHandler
{
    String          appName     = "tibmapget";
    String          realmServer = null;
    String          mapEndpoint = "tibmapget-endpoint";
    String          trcLevel    = null;
    String          user        = "guest";
    String          password    = "guest-pw";
    String          identifier  = null;
    String          secondary   = null;
    String          mapName     = "tibmap";
    static final String DEFAULT_URL = "http://localhost:8080";
    static final String DEFAULT_SECURE_URL = "https://localhost:8080";
    String          keyList     = null;
    String          clientLabel = null;

    Realm           realm       = null;
    TibMap          map         = null;
    boolean         trustAll    = false;
    String          trustFile   = null;
    boolean         removeMap   = false;

    String          usageString =
        "TibMapGet [options] url\n" +
        "Default url is " + TibMapGet.DEFAULT_URL + "\n" +
        "  where options can be:\n" +
        "    -a, --application <name>\n" +
        "    -cl, --client-label <string>\n" +
        "    -e, --endpoint <name>\n" +
        "    -h, --help\n" +
        "    -id, --identifier <string>\n" +
        "    -l, --keyList <string>[,<string>]\n" +
        "    -n, --mapname <name>\n" +
        "    -p, --password <string>\n" +
        "    -r  --removemap \n" +
        "            the mapname command line arg needs to be specified as well\n" +
        "    -s, --secondary <string>\n" +
        "    -t, --trace <level>\n" +
        "           where <level> can be:\n" +
        "                 off, severe, warn, info, debug, verbose\n" +
        "        --trustall\n" +
        "        --trustfile <file>\n" +
        "    -u, --user <string>\n";

    public TibMapGet(String[] args)
    {
        System.out.printf("#\n# %s\n#\n# %s\n#\n",
                          this.getClass().getName(),
                          FTL.getVersionInformation());
        
        parseArgs(args);
    }

    public void usage()
    {
        System.out.println(usageString);
        System.exit(0);
    }

    public void parseArgs(String args[])
    {
        int i           = 0;
        int argc        = args.length;
        String s        = null;
        int n           = 0;
        TibAux aux      = new TibAux(args, usageString);

        for (i = 0; i < argc; i++)
        {
            s = aux.getString(i, "--application", "-a");
            if (s != null) {
                appName = s;
                i++;
                continue;
            }
            s = aux.getString(i, "--client-label", "-cl");
            if (s != null) {
                clientLabel = s;
                i++;
                continue;
            }
            s = aux.getString(i, "--endpoint", "-e");
            if (s != null) {
                mapEndpoint = s;
                i++;
                continue;
            }
            if (aux.getFlag(i, "--help", "-h")) {
                usage();
            }
            s = aux.getString(i, "--identifier", "-id");
            if (s != null) {
                identifier = s;
                i++;
                continue;
            }
            s = aux.getString(i, "--keyList", "-l");
            if (s != null) {
                keyList = s;
                i++;
                continue;
            }
            s = aux.getString(i, "--mapname", "-n");
            if (s != null) {
                mapName = s;
                i++;
                continue;
            }
            s = aux.getString(i, "--password", "-p");
            if (s != null) {
                password = s;
                i++;
                continue;
            }
            if (aux.getFlag(i, "--removemap", "-r")) {
                removeMap = true;
                continue;
            }
            s = aux.getString(i, "--secondary", "-s");
            if (s != null) {
                secondary = s;
                i++;
                continue;
            }
            s = aux.getString(i, "--trace", "-t");
            if (s != null) {
                trcLevel = s;
                i++;
                continue;
            }
            if (aux.getFlag(i, "--trustall", "--trustall")) {
                if (trustFile != null)
                {
                    System.out.println("cannot specify both --trustall and --trustfile");
                    usage();
                }
                trustAll = true;
                continue;
            }
            s = aux.getString(i, "--trustfile", "--trustfile");
            if (s != null)
            {
                trustFile = s;
                if (trustAll)
                {
                    System.out.println("cannot specify both --trustall and --trustfile");
                    usage();
                }
                i++;
                continue;
            }
            s = aux.getString(i, "--user", "-u");
            if (s != null) {
                user = s;
                i++;
                continue;
            }
            if (args[i].startsWith("-", 0)) {
                System.out.println("invalid option: "+args[i]);
                usage();
            }
            realmServer = args[i];
        }

        System.out.print("Invoked as: TibMapGet ");
        for (i = 0; i < argc; i++)
            System.out.print(args[i]+" ");
        System.out.println("");
    }
    
    public void onNotification(int type, String reason)
    {
        if (type == NotificationHandler.CLIENT_DISABLED)
        {
            System.out.println("application administratively disabled: " + reason);
            System.out.println("exiting");
            System.exit(1);
        }
        else
        {
            System.out.println("notification type " + type + ": " + reason);
        }
    }

    public int work() throws FTLException
    {
        TibProperties   props;
        
        // Set global trace to specified level
        if (trcLevel != null)
            FTL.setLogLevel(trcLevel);

        // Get a single realm per process
        props = FTL.createProperties();
        props.set(Realm.PROPERTY_STRING_USERNAME, user);
        props.set(Realm.PROPERTY_STRING_USERPASSWORD, password);
        if (identifier != null)
            props.set(Realm.PROPERTY_STRING_APPINSTANCE_IDENTIFIER, identifier);
        if (secondary != null)
            props.set(Realm.PROPERTY_STRING_SECONDARY_SERVER, secondary);

        if (trustAll == true || trustFile != null)
        {
            if (realmServer == null)
                realmServer = TibMapGet.DEFAULT_SECURE_URL;

            if (trustAll)
                props.set(Realm.PROPERTY_LONG_TRUST_TYPE, Realm.HTTPS_CONNECTION_TRUST_EVERYONE);
            else
            {
                props.set(Realm.PROPERTY_LONG_TRUST_TYPE, Realm.HTTPS_CONNECTION_USE_SPECIFIED_TRUST_FILE);
                props.set(Realm.PROPERTY_STRING_TRUST_FILE, trustFile);
            }
        }
        else
        {
            if (realmServer == null)
                realmServer = TibMapGet.DEFAULT_URL;
        }

        if (clientLabel != null)
            props.set(Realm.PROPERTY_STRING_CLIENT_LABEL, clientLabel);
        else
            props.set(Realm.PROPERTY_STRING_CLIENT_LABEL, this.getClass().getSimpleName());

        realm = FTL.connectToRealmServer(realmServer, appName, props);
        props.destroy();

        // Set this object as the NotificationHandler to receive notifications
        realm.setNotificationHandler(this);
        map = realm.createMap(mapEndpoint, mapName, null);

        if (keyList != null)
        {
            StringTokenizer tok = new StringTokenizer(keyList, ",");
            try
            {
                while(tok.hasMoreElements())
                {
                    String key = (String)tok.nextElement();
                    
                    System.out.println("getting value for key " + key);
                    Message msg = map.get(key);
                    if (msg != null)
                        System.out.println("key = " + key + ", value = " + msg);
                    else
                        System.out.println("currently no key '" + key + "' in the map");
                }
            }
            catch (FTLClientShutdownException ftle)
            {
                // Something unusual happened and the Realm was shutdown. Give the
                // notification handler a chance to run so that we can log if this
                // is an administrative action.
                try
                {
                    Thread.sleep(1000);
                }
                catch (InterruptedException ie)
                {
                    // ignore
                }

                ftle.printStackTrace();
            }
        }
        else
        {
            Message msg = null;
            String  key = null;
            TibMapIterator iter;

            System.out.println("iterating keys is the map");

            try 
            {
                iter = map.createIterator(null);
                while(iter.next())
                {
                    key = iter.currentKey();
                    msg = iter.currentValue();
 
                    System.out.println("key = " + key + " message = " + msg);
                }
                iter.destroy();
            }
            catch (FTLException  e)
            {
                e.printStackTrace(); 
            }
        }

        // Clean up
        map.close();

        // if removeMap is true and mapName is specified then remove the map.
        if ((mapName != null) && removeMap)
        {
            System.out.println("Removing map '" + mapName + "'");
            realm.removeMap(mapEndpoint, mapName, null);
        }

        realm.close();

        return 0;
    }
    
    public static void main(String[] args)
    {
        TibMapGet s  = new TibMapGet(args);
        try
        {
            s.work();
        }
        catch (Throwable e)
        {
            e.printStackTrace();
        }
    }
}
